internal/pkg/cosash/cosash.go

```diff
@@ -0,0 +1,163 @@
+	}
+	if serial != expected {
+		return "", fmt.Errorf("unexpected ack serial from cosash; expected=%d reply=%d", expected, serial)
+	}
+	return parts[1], nil
+}
+
+// NewCosaSh creates a new companion shell process
+func NewCosaSh() (*CosaSh, error) {
+	cmd := exec.Command("/bin/bash")
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Pdeathsig: syscall.SIGTERM,
+	}
+	// This is the channel where we send our commands
+	input, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, err
+	}
+	// stdout and stderr are the same as ours; we are effectively
+	// "co-processing", so we want to get output/errors as they're
+	// printed.
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	cmdin, cmdout, err := os.Pipe()
+	if err != nil {
+		return nil, err
+	}
+	cmd.ExtraFiles = append(cmd.ExtraFiles, cmdout)
+
+	// Start the process
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	replychan := make(chan string)
+	errchan := make(chan error)
+
+	r := &CosaSh{
+		input:         input,
+		cmd:           cmd,
+		replychan:     replychan,
+		errchan:       errchan,
+		preparedBuild: false,
+	}
+
+	// Send a message when the process exits
+	go func() {
+		errchan <- cmd.Wait()
+	}()
+	// Parse the ack serials into a channel
+	go func() {
+		bufr := bufio.NewReader(cmdin)
+		for {
+			reply, err := parseAck(bufr, r.ackserial)
+			if err != nil {
+				// Don't propagate EOF, since we want the process exit status instead.
+				if err == io.EOF {
+					break
+				}
+				errchan <- err
+				break
+			}
+			r.ackserial += 1
+			replychan <- reply
+		}
+	}()
+
+	// Initialize the internal library
+	r.process(". /usr/lib/coreos-assembler/cmdlib.sh\n")
+
+	return r, nil
+}
+
+// write sends content to the shell's stdin, synchronously wait for the reply
+func (r *CosaSh) processWithReply(buf string) (string, error) {
+	// Inject code which writes the serial reply prefix
+	cmd := fmt.Sprintf("echo -n \"%d \" >&3\n", r.ackserial)
+	if _, err := io.WriteString(r.input, cmd); err != nil {
+		return "", err
+	}
+	// Tell the shell to execute the code, which should write the reply to fd 3
+	// which will complete the command.
+	if _, err := io.WriteString(r.input, buf); err != nil {
+		return "", err
+	}
+
+	select {
+	case reply := <-r.replychan:
+		return reply, nil
+	case err := <-r.errchan:
+		return "", err
+	}
+}
+
+func (sh *CosaSh) process(buf string) error {
+	buf = fmt.Sprintf("%s\necho OK >&3\n", buf)
+	r, err := sh.processWithReply(buf)
+	if err != nil {
+		return err
+	}
+	if r != "OK" {
+		return fmt.Errorf("unexpected reply from cosash; expected OK, found %s", r)
+	}
+	return nil
+}
+
+// PrepareBuild prepares for a build
+func (sh *CosaSh) PrepareBuild() error {
+	return sh.process("prepare_build")
+}
+
+// HasPrivileges checks if we can use sudo
+func (sh *CosaSh) HasPrivileges() (bool, error) {
+	 r, err := sh.processWithReply(`
+if has_privileges; then
+  echo true >&3
+else
+  echo false >&3
+fi`)
```

Comment:
So personally I think this is a nifty (and yes, also hacky) demonstration of how we can practically not rewrite all the shell script at once.  

To proxy a conditional like this we need to (well, want to) do something a bit more sophisticated than just fork off a process and collect its exit status.

This code demonstrates the bi-directional IPC with a running shell process; here we're returning a boolean, but we could also e.g. have the shell return JSON which we parse, etc.